torch/lib/c10d/TCPStore.cpp

```diff
@@ -57,10 +56,165 @@ TCPStoreDaemon::~TCPStoreDaemon() {
   closeStopSignal();
 }
 
-void TCPStoreDaemon::join() {
+void BackgroundThread::join() {
   daemonThread_.join();
 }
 
+#ifdef _WIN32
+void BackgroundThread::initStopSignal() {
+  ghStopEvent_ = CreateEvent(NULL, TRUE, FALSE, NULL);
+  if (ghStopEvent_ == NULL) {
+    throw std::runtime_error(
+        "Failed to create the control pipe to start the "
+        "BackgroundThread run");
+  }
+}
+
+void BackgroundThread::closeStopSignal() {
+  CloseHandle(ghStopEvent_);
+}
+
+void BackgroundThread::stop() {
+  SetEvent(ghStopEvent_);
+}
+#else
+void BackgroundThread::initStopSignal() {
+  if (pipe(controlPipeFd_.data()) == -1) {
+    throw std::runtime_error(
+        "Failed to create the control pipe to start the "
+        "BackgroundThread run");
+  }
+}
+
+void BackgroundThread::closeStopSignal() {
+  for (auto fd : controlPipeFd_) {
+    if (fd != -1) {
+      ::close(fd);
+    }
+  }
+}
+
+void BackgroundThread::stop() {
+  if (controlPipeFd_[1] != -1) {
+    // close the write end of the pipe
+    ::close(controlPipeFd_[1]);
+    controlPipeFd_[1] = -1;
+  }
+}
+#endif
+
+// TCPStoreListener class methods
+ListenThread::ListenThread(int listenSocket) : BackgroundThread(listenSocket) {
+  daemonThread_ = std::thread(&ListenThread::run, this);
+}
+
+void ListenThread::addCallback(
+    std::string key,
+    std::function<void(c10::optional<std::string>, c10::optional<std::string>)>
+        cb) {
+  keyToCallbacksLock.lock();
```

Comment:
other part of the code base usually use a `lock_guard`. See:

https://github.com/pytorch/pytorch/blob/2bb58a06ef845ec6adf632415857cfc0c04cad9b/torch/csrc/distributed/rpc/tensorpipe_agent.cpp#L989

Reply:
Thanks! This is much better